src/grpc/infer_handler.h

```diff
@@ -664,11 +665,63 @@ class InferHandlerState {
 
     bool ReceivedNotification() { return received_notification_; }
 
+    // Changes the state of grpc_stream_error_state_ to ERROR_CANCELED,
+    // indicating we have closed the stream and initiated the cancel flow
+    void SetGRPCErrorCancelled()
+    {
+      grpc_stream_error_state_ = Triton_grpc_error_steps::ERROR_CANCELED;
+    }
+    // Returns true ONLY when GRPC_ERROR from CORE is waiting to be processed.
+    bool CheckAndUpdateGRPCError()
+    {
+      if (grpc_stream_error_state_ == Triton_grpc_error_steps::ERROR_WAITING) {
+        // Change the state to ERROR_CANCELED as we have called
+        // HandleCancellation
+        SetGRPCErrorCancelled();
+        return true;
+      }
+      return false;
+    }
+
     bool IsCancelled()
     {
-      return received_notification_ ? ctx_->IsCancelled() : false;
+      return received_notification_
+                 ? (ctx_->IsCancelled() || CheckAndUpdateGRPCError())
+                 : false;
     }
 
+    // Extracts headers from GRPC request and updates state
+    void ExtractStateFromHeaders(InferHandlerStateType* state)
+    {
+      const auto& metadata = state->context_->ctx_->client_metadata();
+      for (const auto& pair : metadata) {
+        auto& key = pair.first;
+        auto& value = pair.second;
+        std::string param_key = std::string(key.begin(), key.end());
+        std::string value_key = std::string(value.begin(), value.end());
+        std::string triton_grpc_error_key = "triton_grpc_error";
+        if (param_key == triton_grpc_error_key) {
```

Comment:
I think you can use something like: `metadata.find(...)`:
https://github.com/triton-inference-server/server/blob/53200091b84f08a5e4921f5073137784570283e9/src/grpc/infer_handler.h#L1466